File: tradingagents/runtime/events.py
"""The Quorum streaming event contract (``contract_version = 1``).

A graph run is exposed to clients (the Flutter desktop UI now, a mobile remote later) as an ordered
stream of :class:`Event` objects. Every event is JSON-serializable and carries a monotonic ``seq`` so
a reconnecting client can resume from a ``Last-Event-ID`` without gaps. The same objects are emitted
by the CLI and the FastAPI sidecar — this module is the single source of truth for their shape.

Design notes:
- One ``Event`` dataclass with a ``data`` payload dict (rather than a class per event type) keeps the
  SSE/serialization layer uniform; the module-level builder functions give type-safe construction.
- ``seq``/``run_id``/``ts`` are stamped by the emitter (see :mod:`tradingagents.runtime.runner` and the
  sidecar's event log), so builders leave ``seq = -1`` until then.
"""

from __future__ import annotations

import time
from dataclasses import dataclass, field
from enum import Enum
from typing import Any

#: Bumped when the wire format changes in a backward-incompatible way. Clients negotiate on it.
CONTRACT_VERSION = 1


class EventType(str, Enum):
    """The closed set of streamed event types. ``str`` mixin → JSON-serializes as its value."""

    RUN_STARTED = "run_started"
    STAGE_STARTED = "stage_started"
    STAGE_DONE = "stage_done"
    AGENT_STARTED = "agent_started"
    AGENT_DONE = "agent_done"
    TOKEN = "token"
    TOOL_CALL = "tool_call"
    REPORT_SECTION_DONE = "report_section_done"
    COST = "cost"
    RUN_DONE = "run_done"
    ERROR = "error"
    HEARTBEAT = "heartbeat"


class Stage(str, Enum):
    """The five pipeline phases, in execution order. Used by ``stage_started``/``stage_done``."""

    ANALYSTS = "analysts"
    RESEARCH_DEBATE = "research_debate"
    TRADER = "trader"
    RISK_DEBATE = "risk_debate"
    PORTFOLIO = "portfolio"


class AgentId(str, Enum):
    """Stable agent identifiers used by ``agent_started``/``agent_done`` and ``token``/``tool_call``.

    These are the wire identifiers (snake_case, stable across releases); human-facing display names
    live in the UI layer, not here.
    """

    MARKET = "market"
    SOCIAL = "social"
    NEWS = "news"
    FUNDAMENTALS = "fundamentals"
    BULL = "bull"
    BEAR = "bear"
    RESEARCH_MANAGER = "research_manager"
    TRADER = "trader"
    AGGRESSIVE = "aggressive"
    NEUTRAL = "neutral"
    CONSERVATIVE = "conservative"
    PORTFOLIO = "portfolio"


@dataclass
class Event:
    """A single streamed event.

    ``seq`` is ``-1`` until an emitter/event-log assigns a monotonic value; ``run_id`` and ``ts`` are
    likewise stamped by the emitter (``ts`` defaults to creation time for standalone use/tests).
    """

    type: EventType
    data: dict[str, Any] = field(default_factory=dict)
    seq: int = -1
    run_id: str | None = None
    ts: float = field(default_factory=time.time)

    def to_dict(self) -> dict[str, Any]:
        """JSON-serializable form. ``type`` is rendered as its string value."""
        return {
            "type": self.type.value,
            "seq": self.seq,
            "run_id": self.run_id,
            "ts": self.ts,
            "data": self.data,
        }


# --- Builder functions: type-safe construction; emitter stamps seq/run_id later. ---

def run_started(ticker: str, trade_date: str, asset_type: str, params: dict[str, Any]) -> Event:
    return Event(EventType.RUN_STARTED, {
        "ticker": ticker, "trade_date": trade_date, "asset_type": asset_type,
        "params": params, "contract_version": CONTRACT_VERSION,
    })


def stage_started(stage: Stage) -> Event:
    return Event(EventType.STAGE_STARTED, {"stage": stage.value})


def stage_done(stage: Stage) -> Event:
    return Event(EventType.STAGE_DONE, {"stage": stage.value})


def agent_started(agent: AgentId) -> Event:
    return Event(EventType.AGENT_STARTED, {"agent": agent.value})


def agent_done(agent: AgentId, *, confidence: float | None = None) -> Event:
    data: dict[str, Any] = {"agent": agent.value}
    if confidence is not None:
        data["confidence"] = confidence
    return Event(EventType.AGENT_DONE, data)


def token(agent: AgentId | str, delta: str) -> Event:
    agent_value = agent.value if isinstance(agent, AgentId) else agent
    return Event(EventType.TOKEN, {"agent": agent_value, "delta": delta})


def tool_call(agent: AgentId | str, tool: str, args_summary: str, status: str = "running") -> Event:
    agent_value = agent.value if isinstance(agent, AgentId) else agent
    return Event(EventType.TOOL_CALL, {
        "agent": agent_value, "tool": tool, "args_summary": args_summary, "status": status,
    })


def report_section_done(section: str, markdown: str, structured: dict[str, Any] | None = None) -> Event:
    return Event(EventType.REPORT_SECTION_DONE, {
        "section": section, "markdown": markdown, "structured": structured,
    })


def cost(llm_calls: int, tool_calls: int, tokens_in: int, tokens_out: int,
         est_usd: float | None = None) -> Event:
    return Event(EventType.COST, {
        "llm_calls": llm_calls, "tool_calls": tool_calls,
        "tokens_in": tokens_in, "tokens_out": tokens_out, "est_usd": est_usd,
    })


def run_done(final_decision: str, *, rating: str | None = None, confidence: float | None = None,
             thesis: str | None = None, structured: dict[str, Any] | None = None) -> Event:
    return Event(EventType.RUN_DONE, {
        "final_decision": final_decision, "rating": rating, "confidence": confidence,
        "thesis": thesis, "structured": structured,
    })


def error(where: str, message: str, *, recoverable: bool = False) -> Event:
    return Event(EventType.ERROR, {"where": where, "message": message, "recoverable": recoverable})


def heartbeat() -> Event:
    return Event(EventType.HEARTBEAT, {})

File: packages/quorum_core/lib/src/events.dart
/// The Quorum streaming event contract in Dart — a sealed union mirroring the Python
/// `tradingagents/runtime/events.py` (contract_version 1).
///
/// Hand-written (no codegen) so the reducer gets Dart 3 exhaustive switches with zero build_runner
/// toolchain. `QuorumEvent.fromEnvelope` parses the wire shape `{seq, run_id, ts, type, data}` and
/// is forward-compatible: an unrecognized `type` becomes [UnknownEvent] rather than throwing.
library;

const int contractVersion = 1;

/// The five pipeline phases, in order. [Stage.unknown] guards forward-compat.
enum Stage { analysts, researchDebate, trader, riskDebate, portfolio, unknown }

/// Stable agent identifiers. [AgentId.unknown] guards forward-compat.
enum AgentId {
  market, social, news, fundamentals,
  bull, bear, researchManager, trader,
  aggressive, neutral, conservative, portfolio,
  unknown,
}

Stage stageFromWire(String? s) => switch (s) {
      'analysts' => Stage.analysts,
      'research_debate' => Stage.researchDebate,
      'trader' => Stage.trader,
      'risk_debate' => Stage.riskDebate,
      'portfolio' => Stage.portfolio,
      _ => Stage.unknown,
    };

AgentId agentFromWire(String? s) => switch (s) {
      'market' => AgentId.market,
      'social' => AgentId.social,
      'news' => AgentId.news,
      'fundamentals' => AgentId.fundamentals,
      'bull' => AgentId.bull,
      'bear' => AgentId.bear,
      'research_manager' => AgentId.researchManager,
      'trader' => AgentId.trader,
      'aggressive' => AgentId.aggressive,
      'neutral' => AgentId.neutral,
      'conservative' => AgentId.conservative,
      'portfolio' => AgentId.portfolio,
      _ => AgentId.unknown,
    };

sealed class QuorumEvent {
  final int seq;
  final String? runId;
  final double ts;
  const QuorumEvent({required this.seq, required this.runId, required this.ts});

  /// Parse a wire envelope `{seq, run_id, ts, type, data}` into a typed event.
  static QuorumEvent fromEnvelope(Map<String, dynamic> env) {
    final seq = (env['seq'] as num?)?.toInt() ?? -1;
    final runId = env['run_id'] as String?;
    final ts = (env['ts'] as num?)?.toDouble() ?? 0;
    final data = ((env['data'] as Map?) ?? const {}).cast<String, dynamic>();
    switch (env['type'] as String?) {
      case 'run_started':
        return RunStarted(seq: seq, runId: runId, ts: ts,
            ticker: data['ticker'] as String? ?? '',
            tradeDate: data['trade_date'] as String? ?? '',
            assetType: data['asset_type'] as String? ?? 'stock');
      case 'stage_started':
        return StageStarted(seq: seq, runId: runId, ts: ts, stage: stageFromWire(data['stage'] as String?));
      case 'stage_done':
        return StageDone(seq: seq, runId: runId, ts: ts, stage: stageFromWire(data['stage'] as String?));
      case 'agent_started':
        return AgentStarted(seq: seq, runId: runId, ts: ts, agent: agentFromWire(data['agent'] as String?));
      case 'agent_done':
        return AgentDone(seq: seq, runId: runId, ts: ts,
            agent: agentFromWire(data['agent'] as String?),
            confidence: (data['confidence'] as num?)?.toDouble());
      case 'token':
        return TokenDelta(seq: seq, runId: runId, ts: ts,
            agent: data['agent'] as String? ?? 'system', delta: data['delta'] as String? ?? '');
      case 'tool_call':
        return ToolCall(seq: seq, runId: runId, ts: ts,
            agent: data['agent'] as String? ?? 'system',
            tool: data['tool'] as String? ?? '',
            argsSummary: data['args_summary'] as String? ?? '',
            status: data['status'] as String? ?? '');
      case 'report_section_done':
        return ReportSectionDone(seq: seq, runId: runId, ts: ts,
            section: data['section'] as String? ?? '',
            markdown: data['markdown'] as String? ?? '',
            structured: (data['structured'] as Map?)?.cast<String, dynamic>());
      case 'cost':
        return CostEvent(seq: seq, runId: runId, ts: ts,
            llmCalls: (data['llm_calls'] as num?)?.toInt() ?? 0,
            toolCalls: (data['tool_calls'] as num?)?.toInt() ?? 0,
            tokensIn: (data['tokens_in'] as num?)?.toInt() ?? 0,
            tokensOut: (data['tokens_out'] as num?)?.toInt() ?? 0,
            estUsd: (data['est_usd'] as num?)?.toDouble());
      case 'run_done':
        return RunDone(seq: seq, runId: runId, ts: ts,
            finalDecision: data['final_decision'] as String? ?? '',
            rating: data['rating'] as String?,
            confidence: (data['confidence'] as num?)?.toDouble(),
            thesis: data['thesis'] as String?,
            structured: (data['structured'] as Map?)?.cast<String, dynamic>(),
            cancelled: data['cancelled'] as bool? ?? false);
      case 'error':
        return ErrorEvent(seq: seq, runId: runId, ts: ts,
            where: data['where'] as String? ?? '',
            message: data['message'] as String? ?? '',
            recoverable: data['recoverable'] as bool? ?? false);
      case 'heartbeat':
        return Heartbeat(seq: seq, runId: runId, ts: ts);
      case final t:
        return UnknownEvent(seq: seq, runId: runId, ts: ts, type: t ?? 'unknown', data: data);
    }
  }
}

final class RunStarted extends QuorumEvent {
  final String ticker, tradeDate, assetType;
  const RunStarted({required super.seq, required super.runId, required super.ts,
      required this.ticker, required this.tradeDate, required this.assetType});
}

final class StageStarted extends QuorumEvent {
  final Stage stage;
  const StageStarted({required super.seq, required super.runId, required super.ts, required this.stage});
}

final class StageDone extends QuorumEvent {
  final Stage stage;
  const StageDone({required super.seq, required super.runId, required super.ts, required this.stage});
}

final class AgentStarted extends QuorumEvent {
  final AgentId agent;
  const AgentStarted({required super.seq, required super.runId, required super.ts, required this.agent});
}

final class AgentDone extends QuorumEvent {
  final AgentId agent;
  final double? confidence;
  const AgentDone({required super.seq, required super.runId, required super.ts, required this.agent, this.confidence});
}

final class TokenDelta extends QuorumEvent {
  final String agent;
  final String delta;
  const TokenDelta({required super.seq, required super.runId, required super.ts, required this.agent, required this.delta});
}

final class ToolCall extends QuorumEvent {
  final String agent, tool, argsSummary, status;
  const ToolCall({required super.seq, required super.runId, required super.ts,
      required this.agent, required this.tool, required this.argsSummary, required this.status});
}

final class ReportSectionDone extends QuorumEvent {
  final String section, markdown;
  final Map<String, dynamic>? structured;
  const ReportSectionDone({required super.seq, required super.runId, required super.ts,
      required this.section, required this.markdown, this.structured});
}

final class CostEvent extends QuorumEvent {
  final int llmCalls, toolCalls, tokensIn, tokensOut;
  final double? estUsd;
  const CostEvent({required super.seq, required super.runId, required super.ts,
      required this.llmCalls, required this.toolCalls, required this.tokensIn, required this.tokensOut, this.estUsd});
}

final class RunDone extends QuorumEvent {
  final String finalDecision;
  final String? rating, thesis;
  final double? confidence;
  final Map<String, dynamic>? structured;
  final bool cancelled;
  const RunDone({required super.seq, required super.runId, required super.ts,
      required this.finalDecision, this.rating, this.confidence, this.thesis, this.structured, this.cancelled = false});
}

final class ErrorEvent extends QuorumEvent {
  final String where, message;
  final bool recoverable;
  const ErrorEvent({required super.seq, required super.runId, required super.ts,
      required this.where, required this.message, this.recoverable = false});
}

final class Heartbeat extends QuorumEvent {
  const Heartbeat({required super.seq, required super.runId, required super.ts});
}

final class UnknownEvent extends QuorumEvent {
  final String type;
  final Map<String, dynamic> data;
  const UnknownEvent({required super.seq, required super.runId, required super.ts, required this.type, required this.data});
}
